File: web/packages/wsh/src/session.mjs
/**
 * WshSession — manages a single PTY or exec channel over a wsh transport.
 *
 * Each session owns a pair of data streams (stdin/stdout) and receives
 * control messages (EXIT, CLOSE, RESIZE) dispatched by the parent WshClient.
 * Data flows as raw bytes on the transport's bidirectional streams, with no
 * CBOR framing overhead.
 */

import {
  MSG, resize as resizeMsg, signal as signalMsg, close as closeMsg,
} from './messages.mjs';

// ── Session states ────────────────────────────────────────────────────

const STATE_OPENING = 'opening';
const STATE_ACTIVE  = 'active';
const STATE_CLOSED  = 'closed';

// ── Text encoding ─────────────────────────────────────────────────────

const textEncoder = new TextEncoder();

// ── Session class ─────────────────────────────────────────────────────

export class WshSession {
  /** @type {string} Unique session identifier (typically matches channelId). */
  id;

  /** @type {'pty'|'exec'} Channel kind. */
  kind;

  /** @type {number} Channel identifier assigned by the server. */
  channelId;

  /** @type {'opening'|'active'|'closed'} Current session state. */
  #state = STATE_OPENING;

  /** @type {import('./transport.mjs').WshTransport} Transport reference. */
  #transport;

  /**
   * Stream IDs returned by the server in OPEN_OK.
   * Typically { stdin: number, stdout: number } or a single bidirectional ID.
   * @type {object}
   */
  #streamIds;

  /**
   * The writable side of the stdin data stream.
   * @type {WritableStreamDefaultWriter|null}
   */
  #stdinWriter = null;

  /**
   * The readable side of the stdout data stream.
   * @type {ReadableStream<Uint8Array>|null}
   */
  #stdoutReadable = null;

  /** @type {AbortController} Cancels the background data pump. */
  #abort = new AbortController();

  /** @type {Promise<void>|null} Resolves when the data pump finishes. */
  #pumpDone = null;

  /** @type {number|null} Exit code received from the server. */
  #exitCode = null;

  // ── Callbacks ───────────────────────────────────────────────────────

  /**
   * Called when stdout/stderr data arrives.
   * @type {function(Uint8Array): void|null}
   */
  onData = null;

  /**
   * Called when the remote process exits.
   * @type {function(number): void|null}
   */
  onExit = null;

  /**
   * Called when the session is fully closed.
   * @type {function(): void|null}
   */
  onClose = null;

  // ── Constructor ─────────────────────────────────────────────────────

  /**
   * @param {import('./transport.mjs').WshTransport} transport
   * @param {number} channelId
   * @param {object} streamIds - Stream identifiers from OPEN_OK.
   * @param {'pty'|'exec'} kind
   */
  constructor(transport, channelId, streamIds, kind) {
    this.#transport = transport;
    this.channelId = channelId;
    this.#streamIds = streamIds;
    this.kind = kind;
    this.id = String(channelId);
  }

  /** Current session state. */
  get state() {
    return this.#state;
  }

  /** Exit code, if the process has exited. */
  get exitCode() {
    return this.#exitCode;
  }

  // ── Stream binding ──────────────────────────────────────────────────

  /**
   * Bind the raw data streams to this session and start the read pump.
   * Called by WshClient after stream setup.
   *
   * @param {ReadableStream<Uint8Array>} readable - stdout/stderr bytes from server
   * @param {WritableStream<Uint8Array>} writable - stdin bytes to server
   */
  _bind(readable, writable) {
    if (this.#state === STATE_CLOSED) {
      throw new Error('Cannot bind streams to a closed session');
    }
    this.#stdoutReadable = readable;
    this.#stdinWriter = writable.getWriter();
    this.#state = STATE_ACTIVE;
    this.#pumpDone = this._pumpDataStream();
  }

  // ── Public API ──────────────────────────────────────────────────────

  /**
   * Write data to the session's stdin stream.
   * Accepts a Uint8Array for raw bytes or a string (UTF-8 encoded).
   *
   * @param {Uint8Array|string} data
   */
  async write(data) {
    if (this.#state === STATE_CLOSED) {
      throw new Error('Cannot write to a closed session');
    }
    if (this.#stdinWriter === null) {
      throw new Error('Session not yet bound — stdin writer unavailable');
    }

    const bytes = typeof data === 'string' ? textEncoder.encode(data) : data;
    await this.#stdinWriter.write(bytes);
  }

  /**
   * Request the remote PTY to resize.
   *
   * @param {number} cols - Terminal columns
   * @param {number} rows - Terminal rows
   */
  async resize(cols, rows) {
    this.#assertNotClosed('resize');
    await this.#transport.sendControl(
      resizeMsg({ channelId: this.channelId, cols, rows })
    );
  }

  /**
   * Send a signal to the remote process (e.g. 'SIGINT', 'SIGTERM').
   *
   * @param {string} sig - Signal name
   */
  async signal(sig) {
    this.#assertNotClosed('signal');
    await this.#transport.sendControl(
      signalMsg({ channelId: this.channelId, signal: sig })
    );
  }

  /**
   * Close this session. Sends a CLOSE control message and tears down
   * the data streams. Safe to call multiple times.
   */
  async close() {
    if (this.#state === STATE_CLOSED) return;

    // Optimistically mark closed so no further writes are accepted.
    this.#state = STATE_CLOSED;
    this.#abort.abort();

    // Send CLOSE to the server (best-effort).
    try {
      await this.#transport.sendControl(
        closeMsg({ channelId: this.channelId })
      );
    } catch {
      // Transport may already be closed; ignore.
    }

    // Release the stdin writer.
    try {
      await this.#stdinWriter?.close();
    } catch {
      // May already be closed or errored.
    }

    // Wait for the data pump to finish.
    if (this.#pumpDone) {
      await this.#pumpDone.catch(() => {});
    }

    this.#stdinWriter = null;
    this.#stdoutReadable = null;
    this.#emitClose();
  }

  // ── Control message dispatch ────────────────────────────────────────

  /**
   * Handle a control message dispatched by WshClient for this channel.
   * @param {object} msg - Decoded CBOR control message
   * @internal
   */
  _handleControlMessage(msg) {
    switch (msg.type) {
      case MSG.EXIT: {
        this.#exitCode = msg.code ?? -1;
        try {
          this.onExit?.(this.#exitCode);
        } catch (err) {
          console.error('[wsh:session] onExit handler error:', err);
        }
        break;
      }

      case MSG.CLOSE: {
        // Server-initiated close.
        if (this.#state !== STATE_CLOSED) {
          this.#state = STATE_CLOSED;
          this.#abort.abort();
          this.#releaseStreams();
          this.#emitClose();
        }
        break;
      }

      case MSG.RESIZE: {
        // Server acknowledgment of resize; currently a no-op on the client.
        break;
      }

      default:
        // Unknown control message for this channel — ignore gracefully.
        break;
    }
  }

  // ── Data stream pump ────────────────────────────────────────────────

  /**
   * Continuously read from the stdout data stream and invoke onData.
   * Runs until the stream ends, errors, or the session is aborted.
   * @private
   */
  async _pumpDataStream() {
    if (!this.#stdoutReadable) return;

    const reader = this.#stdoutReadable.getReader();
    try {
      while (true) {
        if (this.#abort.signal.aborted) break;

        const { done, value } = await reader.read();
        if (done) break;

        if (value && value.byteLength > 0) {
          try {
            this.onData?.(value);
          } catch (err) {
            console.error('[wsh:session] onData handler error:', err);
          }
        }
      }
    } catch (err) {
      // Only report errors if we haven't been intentionally aborted.
      if (!this.#abort.signal.aborted) {
        console.error('[wsh:session] data stream read error:', err);
      }
    } finally {
      reader.releaseLock();
    }

    // If the data stream ended but we haven't received a CLOSE message,
    // transition to closed state.
    if (this.#state !== STATE_CLOSED) {
      this.#state = STATE_CLOSED;
      this.#releaseStreams();
      this.#emitClose();
    }
  }

  // ── Private helpers ─────────────────────────────────────────────────

  /**
   * Release stream resources without sending a CLOSE message.
   * @private
   */
  #releaseStreams() {
    try {
      this.#stdinWriter?.close();
    } catch {
      // Ignore.
    }
    this.#stdinWriter = null;
    this.#stdoutReadable = null;
  }

  /**
   * Emit the onClose callback exactly once.
   * @private
   */
  #emitClose() {
    try {
      this.onClose?.();
    } catch (err) {
      console.error('[wsh:session] onClose handler error:', err);
    }
    // Clear callbacks to prevent repeat invocations.
    this.onClose = null;
  }

  /**
   * Throw if the session is closed.
   * @param {string} action - Description of the attempted action.
   * @private
   */
  #assertNotClosed(action) {
    if (this.#state === STATE_CLOSED) {
      throw new Error(`Cannot ${action}: session is closed`);
    }
  }
}

File: crates/wsh-client/src/session.rs
//! Client-side wsh session.
//!
//! A `WshSession` wraps a data stream and provides read/write/resize/signal/close
//! operations on a single channel.

use std::sync::Arc;
use tokio::sync::Mutex;
use wsh_core::error::{WshError, WshResult};
use wsh_core::messages::ChannelKind;
use wsh_core::transport::ByteStream;

/// The state of a session channel.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SessionState {
    /// The channel is open and active.
    Open,
    /// The channel is closing (sent close, waiting for confirmation).
    Closing,
    /// The channel is fully closed.
    Closed,
}

/// Options for opening a new session.
#[derive(Debug, Clone)]
pub struct SessionOpts {
    /// Channel kind (pty, exec, meta, file).
    pub kind: ChannelKind,
    /// Command to execute (for exec channels).
    pub command: Option<String>,
    /// Terminal columns (for pty channels).
    pub cols: Option<u16>,
    /// Terminal rows (for pty channels).
    pub rows: Option<u16>,
    /// Environment variables to set.
    pub env: Option<std::collections::HashMap<String, String>>,
}

impl Default for SessionOpts {
    fn default() -> Self {
        Self {
            kind: ChannelKind::Pty,
            command: None,
            cols: Some(80),
            rows: Some(24),
            env: None,
        }
    }
}

/// Information about an existing session.
#[derive(Debug, Clone)]
pub struct SessionInfo {
    /// Session identifier.
    pub session_id: String,
    /// Channel ID.
    pub channel_id: u32,
    /// Channel kind.
    pub kind: ChannelKind,
    /// Current state.
    pub state: SessionState,
    /// Human-readable name (if set).
    pub name: Option<String>,
}

/// A client-side wsh session wrapping a data stream.
///
/// Provides buffered read/write operations plus control actions
/// (resize, signal, close) that are dispatched via the control channel.
pub struct WshSession {
    /// Channel ID assigned by the server.
    channel_id: u32,
    /// Channel kind.
    kind: ChannelKind,
    /// Current state.
    state: Arc<Mutex<SessionState>>,
    /// The data stream for this channel.
    stream: Arc<Mutex<Box<dyn ByteStream>>>,
    /// Sender for control messages (resize, signal, close) — sent to the client's
    /// control dispatch loop.
    control_tx: tokio::sync::mpsc::Sender<ControlAction>,
}

/// Internal control actions that the session sends to the client dispatch loop.
#[derive(Debug)]
pub enum ControlAction {
    Resize {
        channel_id: u32,
        cols: u16,
        rows: u16,
    },
    Signal {
        channel_id: u32,
        signal: String,
    },
    Close {
        channel_id: u32,
    },
}

impl WshSession {
    /// Create a new session. Called internally by `WshClient::open_session`.
    pub(crate) fn new(
        channel_id: u32,
        kind: ChannelKind,
        stream: Box<dyn ByteStream>,
        control_tx: tokio::sync::mpsc::Sender<ControlAction>,
    ) -> Self {
        Self {
            channel_id,
            kind,
            state: Arc::new(Mutex::new(SessionState::Open)),
            stream: Arc::new(Mutex::new(stream)),
            control_tx,
        }
    }

    /// The channel ID assigned by the server.
    pub fn channel_id(&self) -> u32 {
        self.channel_id
    }

    /// The kind of this channel.
    pub fn kind(&self) -> &ChannelKind {
        &self.kind
    }

    /// Current session state.
    pub async fn state(&self) -> SessionState {
        *self.state.lock().await
    }

    /// Write data to the session's data stream.
    pub async fn write(&self, data: &[u8]) -> WshResult<()> {
        let state = self.state.lock().await;
        if *state != SessionState::Open {
            return Err(WshError::Channel(format!(
                "channel {} is not open (state: {:?})",
                self.channel_id, *state
            )));
        }
        drop(state);

        let mut stream = self.stream.lock().await;
        stream.write_all(data).await
    }

    /// Read data from the session's data stream.
    ///
    /// Returns the number of bytes read. Returns 0 on EOF.
    pub async fn read(&self, buf: &mut [u8]) -> WshResult<usize> {
        let mut stream = self.stream.lock().await;
        stream.read(buf).await
    }

    /// Resize the terminal (for pty sessions).
    pub async fn resize(&self, cols: u16, rows: u16) -> WshResult<()> {
        self.control_tx
            .send(ControlAction::Resize {
                channel_id: self.channel_id,
                cols,
                rows,
            })
            .await
            .map_err(|_| WshError::Channel("control channel closed".into()))
    }

    /// Send a signal to the session process.
    pub async fn signal(&self, sig: &str) -> WshResult<()> {
        self.control_tx
            .send(ControlAction::Signal {
                channel_id: self.channel_id,
                signal: sig.to_string(),
            })
            .await
            .map_err(|_| WshError::Channel("control channel closed".into()))
    }

    /// Close this session.
    pub async fn close(&self) -> WshResult<()> {
        {
            let mut state = self.state.lock().await;
            if *state == SessionState::Closed {
                return Ok(());
            }
            *state = SessionState::Closing;
        }

        self.control_tx
            .send(ControlAction::Close {
                channel_id: self.channel_id,
            })
            .await
            .map_err(|_| WshError::Channel("control channel closed".into()))?;

        // Close the data stream
        {
            let mut stream = self.stream.lock().await;
            stream.close().await?;
        }

        {
            let mut state = self.state.lock().await;
            *state = SessionState::Closed;
        }

        Ok(())
    }

    /// Mark this session as closed (called externally when an Exit message is received).
    pub(crate) async fn mark_closed(&self) {
        let mut state = self.state.lock().await;
        *state = SessionState::Closed;
    }

    /// Mark this session as open (called when transitioning from a connecting state).
    #[allow(dead_code)]
    pub(crate) async fn mark_open(&self) {
        let mut state = self.state.lock().await;
        *state = SessionState::Open;
    }
}
